### PR TITLE
Promise-ify all tests, and pass a Promise reject function to mocking utilities.

### PR DESCRIPTION
**Pull request by benjamn**
Because Jest has no ability to trace uncaught exceptions back to the test responsible for throwing them, the Apollo Client test suite has been plagued for a long time by "No more mocked responses for the query" errors, often causing unrelated tests to fail.

I have finally tracked the source of these errors down to the implementation of `MockLink`, which was using `throw` to raise exceptions, potentially outside the bounds of the test for which the `MockLink` was originally created.

This gigantic commit does two things:
* Make all `done`/`done.fail`-style tests use `Promise` `resolve` and `reject` functions instead (the bulk of the changes). This is important because we don't want stray `done.fail` calls causing tests that have already passed to fail, and `resolve`/`reject` are conveniently idempotent.
* Make the mocking utilities (`MockLink`, `mockSingleLink`, etc.) require a `Promise` `reject` function, so that they can safely report errors back to the true original test that created the `Promise`.

A better testing framework would maintain an execution context for each test that allowed exceptions to be attributed to the correct test without explicitly passing around `reject` functions, but doing that right probably requires some ECMAScript language support (or at least a transpiler to simulate it). This is an exciting future possibility, but not something we can tackle here.

### CODE REVIEW

File: src/__mocks__/mockLinks.ts [hunk 1]
```diff
@@ -67,18 +74,20 @@ export class MockLink extends ApolloLink {
     const key = requestToKey(operation);
     const responses = this.mockedResponsesByKey[key];
     if (!responses || responses.length === 0) {
-      throw new Error(
+      this.reject(new Error(
```
Comment (benjamn) [inline src/__mocks__/mockLinks.ts]: This is the key to silencing "No more mocked responses for the query" errors after the original test has already passed (or failed).

File: src/__mocks__/mockLinks.ts [hunk 2]
```diff
@@ -45,8 +46,14 @@ export class MockLink extends ApolloLink {
...
   private mockedResponsesByKey: { [key: string]: MockedResponse[] } = {};
 
-  constructor(mockedResponses: MockedResponse[]) {
+  constructor(
+    private reject: (reason: any) => any,
+    mockedResponses: MockedResponse[],
+  ) {
     super();
+    if (typeof this.reject !== "function") {
+      throw new Error("Must pass a failure callback when creating MockLink");
```
Comment (benjamn) [inline src/__mocks__/mockLinks.ts]: Making this an early error was incredibly helpful for tracking down failures to pass in the new `reject` function.

File: src/react/testing/mocks/mockLink.ts [hunk 3]
```diff
@@ -30,10 +30,14 @@ export class MockLink extends ApolloLink {
   private mockedResponsesByKey: { [key: string]: MockedResponse[] } = {};
 
   constructor(
+    private reject: (reason: any) => any,
```
Comment (benjamn) [inline src/react/testing/mocks/mockLink.ts]: @hwillson As we discussed over Slack, I think we should unify the React Apollo versions of `MockLink` etc. with the Apollo Client implementations, now that everything resides in the same codebase.

File: src/react/hooks/__tests__/useMutation.test.tsx [hunk 4]
```diff
@@ -268,7 +268,7 @@ describe('useMutation Hook', () => {
   });
 
   describe('Optimistic response', () => {
-    it('should support optimistic response handling', async () => {
+    it('should support optimistic response handling', async () => new Promise((resolve, reject) => {
```
Comment (benjamn) [inline src/react/hooks/__tests__/useMutation.test.tsx]: I would love to shorten this to
```suggestion
    it.async('should support optimistic response handling', async (resolve, reject) => {
```
I will push a follow-up commit to this PR if I can figure out a good way to add that `it.async` API.

File: src/__tests__/optimistic.ts [hunk 5]
```diff
@@ -188,9 +192,9 @@ describe('optimistic mutation results', () => {
         },
       };
 
-      it('handles a single error for a single mutation', async () => {
+      it('handles a single error for a single mutation', () => new Promise(async (resolve, reject) => {
```
Comment (hwillson) [inline src/__tests__/optimistic.ts]: It's super unfortunate we have to use a Promise anti-pattern to work around Jest's limitations 🙁.

File: src/core/__tests__/QueryManager/index.ts [hunk 6]
```diff
@@ -407,23 +415,23 @@ describe('QueryManager', () => {
...
       console.error = oldError;
-      done();
+      resolve();
     }, 10);
   });
 
   // XXX this looks like a bug in zen-observable but we should figure
   // out a solution for it
-  xit('handles an unsubscribe action that happens before data returns', done => {
+  itAsync.skip('handles an unsubscribe action that happens before data returns', (resolve, reject) => {
```
Comment (benjamn) [inline src/core/__tests__/QueryManager/index.ts]: Pretty cool that `itAsync.skip` works just like `it.skip`!

Comment (hwillson) [reply src/core/__tests__/QueryManager/index.ts]: Love it!

File: src/__tests__/optimistic.ts (hunk 5 above)
Comment (hwillson) [reply src/__tests__/optimistic.ts]: My comment is kinda moot after the changes in https://github.com/apollographql/apollo-client/pull/5474/commits/81e6872684ef83e113a95a87f5d75a7f5271e60d, so resolving.
